File: stack/calculator/1597. Build Binary Expression Tree From Infix Expression/js/main.js
// Time O(N)
// Space O(N)
const expTree = s => {
  let opStack = [];
  let map = { '+': 0, '-': 0, '*': 1, '/': 1, '(': -1 };
  let nodes = [];

  for (let i = 0; i < s.length; i++) {
    if (s[i] == ' ') continue;

    if (s[i] == '(') {
      opStack.push('(');
    } else if (s[i] == ')') {
      // apply operations while prev opStack != '('
      while (opStack[opStack.length - 1] != '(') {
        let newNode = applyOperator(opStack.pop(), nodes.pop(), nodes.pop());
        nodes.push(newNode);
      }

      opStack.pop();
    } else if (s[i] in map) {
      while (opStack.length && map[opStack[opStack.length - 1]] >= map[s[i]]) {
        let newNode = applyOperator(opStack.pop(), nodes.pop(), nodes.pop());
        nodes.push(newNode);
      }

      opStack.push(s[i]);
    } else {
      let buffer = '';
      while (i < s.length && s[i] >= 0 && s[i] <= 9) {
        buffer += s[i++];
      }
      nodes.push(new Node(buffer));
      i--;
    }
  }

  while (opStack.length) {
    let newNode = applyOperator(opStack.pop(), nodes.pop(), nodes.pop());
    nodes.push(newNode);
  }

  return nodes.pop();

  function Node(val, left, right) {
    this.val = val === undefined ? ' ' : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }

  function applyOperator(oper, node1, node2) {
    let node = new Node(oper);
    node.right = node1;
    node.left = node2;
    return node;
  }
};

File: stack/calculator/1597. Build Binary Expression Tree From Infix Expression/py/main.py

# Definition for a binary tree node.
# class Node(object):
#     def __init__(self, val=" ", left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Time O(N)
# Space O(N)
class Solution:
    def expTree(self, s: str) -> 'Node':
      maps = {'+': 0,'-': 0,'*': 1,'/': 1, '(': -1}
      opStack = []
      nodes = []
      i = 0

      while i < len(s):
        if s[i] == ' ': 
          i += 1
          continue
        
        if s[i] == '(':
          opStack.append('(')
        
        elif s[i] == ')':
          while opStack[-1] != '(':
            newNode = self.applyOperator(opStack.pop(), nodes.pop(), nodes.pop())
            nodes.append(newNode)
          opStack.pop()

        elif s[i] in maps:
          while len(opStack) and maps[opStack[-1]] >= maps[s[i]]:
            newNode = self.applyOperator(opStack.pop(), nodes.pop(), nodes.pop())
            nodes.append(newNode)
          opStack.append(s[i])
            
        else:
          buf = ''
          while i < len(s) and s[i].isdigit():
            buf += s[i]
            i += 1      
          i -= 1
          newNode = Node(buf) 
          nodes.append(newNode)
        
        i += 1

      while len(opStack):
        newNode = self.applyOperator(opStack.pop(), nodes.pop(), nodes.pop())
        nodes.append(newNode)

      return nodes.pop()

    def applyOperator(self, oper, node1, node2):
      newNode = Node(oper)
      newNode.left = node2
      newNode.right = node1
      return newNode

